device/src/state_machines/mod.rs

```diff
@@ -2,18 +2,27 @@ use super::*;
 use lorawan_encoding::parser::DecryptedDataPayload;
 
 pub mod no_session;
-
 pub mod session;
 
 pub struct Shared<R: radio::PhyRxTx + Timings> {
     radio: R,
     credentials: Credentials,
-    region: Region,
+    region: region::Configuration,
     mac: Mac,
     // TODO: do something nicer for randomness
     get_random: fn() -> u32,
     buffer: Vec<u8, U256>,
-    data_downlink: Option<DecryptedDataPayload<Vec<u8, U256>>>,
+    downlink: Option<Downlink>,
+}
+
+enum Downlink {
+    Data(DecryptedDataPayload<Vec<u8, U256>>),
+    Join(JoinAccept),
+}
+
+#[derive(Debug)]
+pub struct JoinAccept {
+    pub cflist: Option<[u32; 5]>,
```

Comment:
I'm trying to provide visibility to the user for debug. I think I tried using the type from `lorawan_encoding` but it felt burdensome because of the crypto.